File: crates/runtime/src/builtin/timeout.rs
use crate::internal::{
    format_caught_error,
    runtime_services::{RuntimeServices, run_process_turn_checkpoint},
};
use rquickjs::function::Args;
use rquickjs::{CatchResultExt, Ctx, Persistent, Value};

// Native functions for the timeout implementation
#[rquickjs::module]
pub mod native_module {
    use crate::internal::runtime_services::RuntimeServices;
    use futures::future::abortable;
    use rquickjs::{Ctx, Persistent, Value};
    use std::sync::atomic::Ordering;

    #[rquickjs::function]
    pub fn schedule(
        ctx: Ctx<'_>,
        code_or_fn: Persistent<Value<'static>>,
        delay: u32,
        periodic: bool,
        args: Persistent<Vec<Value<'static>>>,
    ) -> usize {
        let services = ctx
            .userdata::<RuntimeServices>()
            .expect("runtime services not initialized");
        let key = services
            .timers
            .last_abort_id
            .fetch_add(1, Ordering::Relaxed);

        let (task, abort_handle) = abortable(super::scheduled_task(
            ctx.clone(),
            code_or_fn,
            delay,
            periodic,
            args,
            key,
        ));
        services
            .timers
            .abort_handles
            .borrow_mut()
            .insert(key, abort_handle);
        drop(services);
        let task_ctx = ctx.clone();
        ctx.spawn(async move {
            let _ = task.await;
            // Clean up after the task completes naturally
            let services = task_ctx
                .userdata::<RuntimeServices>()
                .expect("runtime services not initialized");
            services.timers.abort_handles.borrow_mut().remove(&key);
            services.timers.unrefed_timers.borrow_mut().remove(&key);
        });
        key
    }

    #[rquickjs::function]
    pub fn clear_schedule(ctx: Ctx<'_>, timeout_id: usize) {
        let services = ctx
            .userdata::<RuntimeServices>()
            .expect("runtime services not initialized");
        services
            .timers
            .unrefed_timers
            .borrow_mut()
            .remove(&timeout_id);
        let mut abort_handles = services.timers.abort_handles.borrow_mut();
        if let Some(handle) = abort_handles.remove(&timeout_id) {
            handle.abort();
        }
    }

    #[rquickjs::function]
    pub fn unref_schedule(ctx: Ctx<'_>, timeout_id: usize) {
        let services = ctx
            .userdata::<RuntimeServices>()
            .expect("runtime services not initialized");
        services
            .timers
            .unrefed_timers
            .borrow_mut()
            .insert(timeout_id);
    }

    #[rquickjs::function]
    pub fn ref_schedule(ctx: Ctx<'_>, timeout_id: usize) {
        let services = ctx
            .userdata::<RuntimeServices>()
            .expect("runtime services not initialized");
        services
            .timers
            .unrefed_timers
            .borrow_mut()
            .remove(&timeout_id);
    }

    #[rquickjs::function]
    pub fn ref_timer_count(ctx: Ctx<'_>) -> usize {
        let services = ctx
            .userdata::<RuntimeServices>()
            .expect("runtime services not initialized");
        let total = services.timers.abort_handles.borrow().len();
        let unrefed = services.timers.unrefed_timers.borrow().len();
        total.saturating_sub(unrefed)
    }
}

// JS functions for the timeout implementation
pub const TIMEOUT_JS: &str = include_str!("timeout.js");

// JS code wiring the timeout module into the global context
pub const WIRE_JS: &str = r#"
        import * as __wasm_rquickjs_timeout from '__wasm_rquickjs_builtin/timeout';
        globalThis.setTimeout = __wasm_rquickjs_timeout.setTimeout;
        globalThis.setImmediate = __wasm_rquickjs_timeout.setImmediate;
        globalThis.setInterval = __wasm_rquickjs_timeout.setInterval;
        globalThis.clearTimeout = __wasm_rquickjs_timeout.clearTimeout;
        globalThis.clearInterval = __wasm_rquickjs_timeout.clearInterval;
        globalThis.clearImmediate = __wasm_rquickjs_timeout.clearImmediate;
        globalThis.__wasm_rquickjs_ref_timer_count = __wasm_rquickjs_timeout.getRefTimerCount;
    "#;

async fn scheduled_task(
    ctx: Ctx<'_>,
    code_or_fn: Persistent<Value<'static>>,
    delay: u32,
    periodic: bool,
    args: Persistent<Vec<Value<'static>>>,
    timer_key: usize,
) {
    #[cfg(feature = "p2")]
    let duration = wstd::time::Duration::from_millis(delay as u64);

    #[cfg(feature = "p3")]
    let duration_ns = (delay as u64).saturating_mul(1_000_000);

    loop {
        #[cfg(feature = "p2")]
        wstd::task::sleep(duration).await;

        #[cfg(feature = "p3")]
        wasip3::clocks::monotonic_clock::wait_for(duration_ns).await;

        run_scheduled_task(ctx.clone(), code_or_fn.clone(), args.clone(), timer_key)
            .catch(&ctx)
            .unwrap_or_else(|e| {
                eprintln!(
                    "Timer callback error escaped JS uncaught handler: {}",
                    format_caught_error(e)
                )
            });

        if !periodic {
            break;
        }

        // Check if the timer was cancelled during the callback
        let services = ctx
            .userdata::<RuntimeServices>()
            .expect("runtime services not initialized");
        if !services
            .timers
            .abort_handles
            .borrow()
            .contains_key(&timer_key)
        {
            break;
        }
    }
}

fn run_scheduled_task(
    ctx: Ctx,
    code_or_fn: Persistent<Value<'static>>,
    args: Persistent<Vec<Value<'static>>>,
    timer_key: usize,
) -> rquickjs::Result<()> {
    // A timer is the next host callback. Give nextTick and Promise jobs from
    // the preceding turn their normal handler opportunity, then report any
    // remaining unhandled rejections before this callback runs.
    run_process_turn_checkpoint(&ctx)?;

    // A rejection listener may cancel the timer whose callback was about to
    // run. The abort signal is observed only at the next async poll, so honor
    // the removed handle synchronously at this callback boundary as Node does.
    if !ctx
        .userdata::<RuntimeServices>()
        .expect("runtime services not initialized")
        .timers
        .abort_handles
        .borrow()
        .contains_key(&timer_key)
    {
        return Ok(());
    }

    let restored_code_or_fn = code_or_fn.restore(&ctx)?;
    let restored_args = args.restore(&ctx)?;

    let result = if let Some(func) = restored_code_or_fn.as_function() {
        let mut args = Args::new(ctx.clone(), restored_args.len());
        args.push_args(&restored_args)?;
        func.call_arg(args)
    } else if let Some(code) = restored_code_or_fn.as_string() {
        ctx.eval(code.to_string()?)
    } else {
        eprintln!("Unsupported value passed to setTimeout or setInterval: {restored_code_or_fn:?}");
        Ok(())
    };

    result?;

    // A timer callback is itself a Node turn. Drain nextTick before Promise jobs and process
    // rejection events to a fixpoint before another timer callback can run.
    run_process_turn_checkpoint(&ctx)?;

    Ok(())
}

File: crates/runtime/src/builtin/timeout.js
import * as timeoutNative from '__wasm_rquickjs_builtin/timeout_native'
import { _captureContext, _restoreContext } from 'node:async_hooks'

class Timeout {
    constructor(id, callback, delay, args, isInterval) {
        this._id = id;
        this._destroyed = false;
        this._refed = true;
        this._callback = callback;
        this._delay = delay;
        this._args = args;
        this._isInterval = isInterval;
        this.__idleTimeout = delay;
        this.__onTimeout = callback;
        this._repeat = isInterval ? delay : null;
    }

    get _idleTimeout() {
        return this.__idleTimeout;
    }

    set _idleTimeout(value) {
        this.__idleTimeout = value;
        if (value === -1) {
            this.close();
        }
    }

    get _onTimeout() {
        return this.__onTimeout;
    }

    set _onTimeout(value) {
        this.__onTimeout = value;
        if (value === null) {
            this.close();
        }
    }

    ref() {
        this._refed = true;
        timeoutNative.ref_schedule(this._id);
        return this;
    }

    unref() {
        this._refed = false;
        timeoutNative.unref_schedule(this._id);
        return this;
    }

    hasRef() {
        return this._refed && !this._destroyed;
    }

    refresh() {
        if (!this._destroyed) {
            timeoutNative.clear_schedule(this._id);
            const bound = this._bound || this._callback.bind(this);
            this._bound = bound;
            this._id = timeoutNative.schedule(bound, this._delay, this._isInterval, this._args);
        }
        return this;
    }

    close() {
        if (!this._destroyed) {
            this._destroyed = true;
            timeoutNative.clear_schedule(this._id);
        }
        return this;
    }

    [Symbol.toPrimitive]() {
        return this._id;
    }

    [Symbol.dispose]() {
        this.close();
    }
}

function validateCallback(callback) {
    if (typeof callback !== 'function') {
        const err = new TypeError('The "callback" argument must be of type function. Received ' + (callback === null ? 'null' : typeof callback));
        err.code = 'ERR_INVALID_ARG_TYPE';
        throw err;
    }
}

const TIMEOUT_MAX = 2 ** 31 - 1;

function normalizeTimerDelay(time) {
    const delay = +time;
    if (delay > TIMEOUT_MAX) {
        if (typeof process !== 'undefined' && typeof process.emitWarning === 'function') {
            process.emitWarning(`${time} does not fit into a 32-bit signed integer.\nTimeout duration was set to 1.`, 'TimeoutOverflowWarning');
        }
        return 1;
    }
    if (!(delay >= 1)) {
        return 0;
    }
    return Math.trunc(delay);
}

function scheduleTimeout(callback, time, args, isInterval) {
    const snapshot = _captureContext();
    const wrapped = function(...a) {
        const currentId = this._id;
        try {
            // Drain pending nextTick callbacks before executing timer callbacks,
            // matching Node.js's guarantee that process.nextTick always fires
            // before timers (setTimeout/setImmediate).
            if (globalThis.__wasm_rquickjs_drainNextTick) {
                globalThis.__wasm_rquickjs_drainNextTick();
            }
            return _restoreContext(snapshot, callback, this, a);
        } catch (e) {
            if (globalThis.__wasm_rquickjs_handleUncaughtError) {
                globalThis.__wasm_rquickjs_handleUncaughtError(e);
            } else if (typeof console !== 'undefined') {
                console.error(e);
            }
        } finally {
            // Support converting setTimeout to interval via _repeat
            if (!this._destroyed && this._id === currentId && !isInterval && this._repeat > 0) {
                const nextId = timeoutNative.schedule(this._bound, this._repeat, false, args);
                this._id = nextId;
                if (!this._refed) {
                    timeoutNative.unref_schedule(nextId);
                }
            }
        }
    };
    const timeout = new Timeout(0, wrapped, time, args, isInterval);
    const bound = wrapped.bind(timeout);
    const id = timeoutNative.schedule(bound, time, isInterval, args);
    timeout._id = id;
    timeout._bound = bound;
    return timeout;
}

export function setTimeout(callback, time, ...args) {
    validateCallback(callback);
    return scheduleTimeout(callback, normalizeTimerDelay(time), args, false);
}

export function setInterval(callback, time, ...args) {
    validateCallback(callback);
    return scheduleTimeout(callback, normalizeTimerDelay(time), args, true);
}

export function setImmediate(callback, ...args) {
    validateCallback(callback);
    return scheduleTimeout(callback, 0, args, false);
}

export function clearTimeout(id) {
    if (id == null) return;
    if (id instanceof Timeout) {
        id.close();
        return;
    }
    if (typeof id === 'number' || typeof id === 'string') {
        const numId = +id;
        if (numId >= 0 && Number.isFinite(numId)) {
            timeoutNative.clear_schedule(numId);
        }
    }
}

export const clearInterval = clearTimeout;
export const clearImmediate = clearTimeout;

export function getRefTimerCount() {
    return timeoutNative.ref_timer_count();
}
